### DISCUSSION TITLE
virtual env can't be used in docker with ExternalPythonOperator

### DISCUSSION THREAD

**Discussion by yeangPan** [apache/airflow#34039] [General] [answered]
Goal: create and virtual env outside docker, use it in docker with ExternalPythonOperator (reused env)

I deployed airflow using official airflow compose file. after creating a virtual env named airflow on host(whose path was mounted in docker), I tested a dag :
```

from airflow import DAG
from airflow.operators.python import ExternalPythonOperator
import pendulum

# os.environ["PATH"] = os.getenv("PATH") + ":/usr/local/airflow/.local/bin"

def virtualenv_fn():
    import sys
    print(sys.executable)

with DAG(dag_id="test_dag", 
         catchup=False,
         schedule="0 0 * * *",
         start_date=pendulum.datetime(2021, 1, 1, tz="UTC")) as dag:
    virtualenv_task = ExternalPythonOperator(
        task_id="test_task",
        # queue="basedata",
        python="/opt/airflow/envs/airflow/bin/python",
        python_callable=virtualenv_fn,
        # requirements=["boto3>=1.17.43"],
        # system_site_packages=False,
        dag=dag,
    )
    
```

I got error msg:
```
[2023-09-02, 18:12:52 UTC] {taskinstance.py:1660} INFO - Exporting env vars: AIRFLOW_CTX_DAG_OWNER='***' AIRFLOW_CTX_DAG_ID='test_dag' AIRFLOW_CTX_TASK_ID='test_task' AIRFLOW_CTX_EXECUTION_DATE='2023-09-01T00:00:00+00:00' AIRFLOW_CTX_TRY_NUMBER='10' AIRFLOW_CTX_DAG_RUN_ID='scheduled__2023-09-01T00:00:00+00:00'
[2023-09-02, 18:12:52 UTC] {python.py:772} WARNING - When checking for Airflow installed in venv got [Errno 2] No such file or directory: '/opt/airflow/envs/airflow/bin/python'
[2023-09-02, 18:12:52 UTC] {python.py:773} WARNING - This means that Airflow is not properly installed by  /opt/***/envs/***/bin/python. Airflow context keys will not be available. Please Install Airflow 2.7.0 in your environment to access them.
[2023-09-02, 18:12:52 UTC] {taskinstance.py:1943} ERROR - Task failed with exception
Traceback (most recent call last):
  File "/home/airflow/.local/lib/python3.11/site-packages/airflow/operators/python.py", line 395, in execute
    return super().execute(context=serializable_context)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/airflow/.local/lib/python3.11/site-packages/airflow/operators/python.py", line 192, in execute
    return_value = self.execute_callable()
                   ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/airflow/.local/lib/python3.11/site-packages/airflow/operators/python.py", line 702, in execute_callable
    raise ValueError(f"Python Path '{python_path}' must exists")
ValueError: Python Path '/opt/***/envs/***/bin/python' must exists
[2023-09-02, 18:12:52 UTC] {taskinstance.py:1400} INFO - Marking task as FAILED. dag_id=test_dag, task_id=test_task, execution_date=20230901T000000, start_date=20230902T181252, end_date=20230902T181252
[2023-09-02, 18:12:52 UTC] {standard_task_runner.py:104} ERROR - Failed to execute job 16 for task test_task (Python Path '/opt/***/envs/***/bin/python' must exists; 1257)
[2023-09-02, 18:12:52 UTC] {local_task_job_runner.py:228} INFO - Task exited with return code 1
[2023-09-02, 18:12:52 UTC] {taskinstance.py:2784} INFO - 0 downstream tasks scheduled from follow-on schedule check
```

after checking python path , I found that:
![image](https://github.com/apache/airflow/assets/21076543/1e865527-56c7-4303-afe5-d94eb2e241d3)
all python symbolic links were broken. 

what's the best practises for the problem, many thanks

**Comment by Taragolis** [comment, marked as answer]
Best Practice: Create virtual env inside of docker
---
```dockerfile
# Simple Dockerfile content
# Replace slim-2.7.0-python3.9 by appropriate tag
# For more info, please have a look at: https://airflow.apache.org/docs/docker-stack/build.html
FROM apache/airflow:slim-2.7.0-python3.9

USER airflow

RUN unset PIP_USER \
  && python3 -m venv /opt/airflow/venv1 \
  # Provide your own requirements
  && /opt/airflow/venv1/bin/python -m pip install requests
```

```bash
docker build . -t awesome-airflow:awesome-tag
```
```
echo "AIRFLOW_IMAGE_NAME=awesome-airflow:awesome-tag" >> .env
```

If you really want to achieve initial goal
---
> [!IMPORTANT]  
> This advise only as sample, do not repeat at home

You need to create your `venv` with `--copies` flag, but I've bet that you have other issues, like permissions, and some packages still have symlinks and other things including Segmentation Faults

> [!WARNING]  
> Really, better create it within docker container, rather than mount venv from host

```console
❯ python -m venv --copies dont_do_that
❯ ls -lah dont_do_that/bin
total 11M
drwxr-xr-x 12 taragolis  384 Sep  2 23:05 .
drwxr-xr-x  6 taragolis  192 Sep  2 23:05 ..
-rw-r--r--  1 taragolis 8.7K Sep  2 23:05 Activate.ps1
-rw-r--r--  1 taragolis 1.9K Sep  2 23:05 activate
-rw-r--r--  1 taragolis  884 Sep  2 23:05 activate.csh
-rw-r--r--  1 taragolis 2.0K Sep  2 23:05 activate.fish
-rwxr-xr-x  1 taragolis  268 Sep  2 23:05 pip
-rwxr-xr-x  1 taragolis  268 Sep  2 23:05 pip3
-rwxr-xr-x  1 taragolis  268 Sep  2 23:05 pip3.9
-rwxr-xr-x  1 taragolis 3.4M Sep  2 23:05 python
-rwxr-xr-x  1 taragolis 3.4M Sep  2 23:05 python3
-rwxr-xr-x  1 taragolis 3.4M Sep  2 23:05 python3.9
```

**Reply by yeangPan** [reply]
Thanks for your nice and patient reply. After hours of exploring, it was indeed a tough way to realize the initial goal. the goal was based on th e reality that I has an ETL management system, provided to ETL users to develop plugin codes(the real ETL codes). through isolating the env from the system, I can easily add a new ETL process(maybe a new python version and totally different pkgs) without restarting the airflow service. The only things I need to do was to create the env outside docker (rather than change docker compose file and init a production process to change it, the action may have consequences on prod system)/ designated env name and submitting the requirments file to the system.

may I ask questions about how to achieve the target above? Or whether it's possible to achieve that goal?

**Comment by potiuk** [comment]
BTW. There is an upcoming Airflow Summit 2023  talk about someone who has done something similar: https://airflowsummit.org/sessions/2023/circumventing-airflows-limitations-around-multitenancy/ (you can watch it live in Toronto but recording will be published some time after).

This is not something supported out-of-the-box by Python and Docker (and by extension by Airflow) but if you spend enough effort on implementing it, you can achieve it - however it is totally custom approach and rather complex to pull of especially if you want to make it work for several teams and generally have it at "Product" level.

Simply speaking Docker and Docker Compose are NOT foreseen to share virtualenvs of Python with host. What you try to do is somethign of a hack that docker and docker compose does not support  (and by extension Airflow cannot support it either until the support for it is implemented). I guess that's not even going to be Docker at this stage when it will work. Python community is working on some other solutions and in a few years I think we will have different ways of sharing multiple virtualenvs across containers.

Until then - there are no "best practices" other than those that @Taragolis described - i.e. building your venvs inside docker  -not mapping them from the host - unless of course you are willing to go the hard way and do your own custom approach that involves a lot of hacks and workarounds.

**Reply by potiuk** [reply]
BTW. One of the big limitation which is currently not solved is not the symlinks. The symlinks are EASY to solve. The big problem is compatibility and potentialy lack of system (not python) libraries for packages that are using system libraries (compiled code mainly but also some standard system libraries. In order to be able to share venv betwen container and host, you would have to make sure that the system librarires are identical in both. And this is very often not the case becasue the container might use different operating systems, different set and versions of libraries and even they can have different set of architectures involved (I..e having host venv on MacOS and trying to use it in Linuux containers in docker is completely impossible)

So what you are trying to achieve is completely not possible if you have any differences betweeen container and host on the system level. So even if your solution will work - it has to be very strictly controlled environment - your host will generally have to have exactly the same OS, the same version of system libraries but also any other libraries that are needed by your venv to run.

So even if you achieve it, you are probably not going to achieve your goal @yeangPan - unless you will strictly control the system/OS installation of whoever is using your solution.

So generally - you are tryint to go into a very windy and dangerous journey if you want to achieve what you say you want, so ... good luck.

**Reply by yeangPan** [reply]
Thanks for your nice and patient reply. then the goal seems to be more unrealistic. what confuses me is that the env created outside docker can be used in docker through terminal, yet can't be recogized by airflow.
![image](https://github.com/apache/airflow/assets/21076543/6f0dfb6e-4fb6-49dd-a4eb-04fa402aecc6)

anyway, I decided to use the way mentioned by @Taragolis, thank both of you guys for your kind and helpful reply, It helps a lot.

**Reply by potiuk** [reply]
It **SEEMS** to be working. But almost for sure it won't.
